src/utils/xhr-utils.js

```diff
@@ -0,0 +1,626 @@
+
+/** @private @const {!Array<string>} */
+const allowedMethods_ = ['GET', 'POST'];
+
+/** @private @const {string} */
+const ALLOW_SOURCE_ORIGIN_HEADER = 'AMP-Access-Control-Allow-Source-Origin';
+
+/**
+ * The "init" argument of the Fetch API. Currently, only "credentials: include"
+ * is implemented.  Note ampCors with explicit false indicates that
+ * __amp_source_origin should not be appended to the URL to allow for
+ * potential caching or response across pages.
+ *
+ * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+ *
+ * @typedef {{
+ *   body: (!Object|!Array|undefined|string),
+ *   credentials: (string|undefined),
+ *   headers: (!JsonObject|undefined),
+ *   method: (string|undefined),
+ *   requireAmpResponseSourceOrigin: (boolean|undefined),
+ *   ampCors: (boolean|undefined)
+ * }}
+ */
+export let FetchInitDef;
+
+/**
+ * Serializes a fetch request so that it can be passed to `postMessage()`,
+ * i.e., can be cloned using the
+ * [structured clone algorithm](http://mdn.io/Structured_clone_algorithm).
+ *
+ * The request is serialized in the following way:
+ *
+ * 1. If the `init.body` is a `FormData`, set content-type header to
+ * `multipart/form-data` and transform `init.body` into an
+ * `!Array<!Array<string>>` holding the list of form entries, where each
+ * element in the array is a form entry (key-value pair) represented as a
+ * 2-element array.
+ *
+ * 2. Return a new object having properties `input` and the transformed
+ * `init`.
+ *
+ * The serialized request is assumed to be de-serialized in the following way:
+ *
+ * 1.If content-type header starts with `multipart/form-data`
+ * (case-insensitive), transform the entry array in `init.body` into a
+ * `FormData` object.
+ *
+ * 2. Pass `input` and transformed `init` to `fetch` (or the constructor of
+ * `Request`).
+ *
+ * Currently only `FormData` used in `init.body` is handled as it's the only
+ * type being used in AMP runtime that needs serialization. The `Headers` type
+ * also needs serialization, but callers should not be passing `Headers`
+ * object in `init`, as that fails `fetchPolyfill` on browsers that don't
+ * support fetch. Some serialization-needing types for `init.body` such as
+ * `ArrayBuffer` and `Blob` are already supported by the structured clone
+ * algorithm. Other serialization-needing types such as `URLSearchParams`
+ * (which is not supported in IE and Safari) and `FederatedCredentials` are
+ * not used in AMP runtime.
+ *
+ * @param {string} input The URL of the XHR to convert to structured
+ *     cloneable.
+ * @param {!FetchInitDef} init The options of the XHR to convert to structured
+ *     cloneable.
+ * @return {{input: string, init: !FetchInitDef}} The serialized structurally-
+ *     cloneable request.
+ * @private
+ */
+export function toStructuredCloneable(input, init) {
+  const newInit = Object.assign({}, init);
+  if (isFormDataWrapper(init.body)) {
+    newInit.headers = newInit.headers || {};
```

Comment:
Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).